libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSource.kt

```diff
@@ -383,51 +423,62 @@ class MapboxNavigationViewportDataSource(
+        }
 
-                val currentLegPoints = completeRoutePoints[currentLegProgress.legIndex]
-                val remainingStepsAfterCurrentStep =
-                    if (currentStepProgress.stepIndex < currentLegPoints.size) {
-                        currentLegPoints.slice(
-                            currentStepProgress.stepIndex + 1 until currentLegPoints.size - 1
-                        )
-                    } else {
-                        emptyList()
-                    }
-                val remainingPointsAfterCurrentStep = remainingStepsAfterCurrentStep.flatten()
-                remainingPointsOnRoute = listOf(
-                    remainingPointsOnCurrentStep,
-                    remainingPointsAfterCurrentStep
-                ).flatten()
-
-                updateData()
-                return
+        ifNonNull(
+            routeProgress.currentLegProgress,
+            routeProgress.currentLegProgress?.currentStepProgress
+        ) { currentLegProgress, currentStepProgress ->
+            options.followingFrameOptions.run {
+                followingPitchProperty.fallback = getPitchFallbackFromRouteProgress(
+                    pitchNearManeuvers.enabled,
+                    pitchNearManeuvers.triggerDistanceFromManeuver,
+                    defaultPitch,
+                    currentStepProgress.distanceRemaining
+                )
+            }
+
+            options.followingFrameOptions.intersectionDensityCalculation.run {
+                pointsToFrameOnCurrentStep = getPointsToFrameOnCurrentStep(
+                    intersectionDensityCalculationEnabled = enabled,
+                    intersectionDensityAverageDistanceMultiplier = this.averageDistanceMultiplier,
+                    averageIntersectionDistancesOnRoute = averageIntersectionDistancesOnRoute,
+                    currentLegProgress = currentLegProgress,
+                    currentStepProgress = currentStepProgress
+                )
             }
+
+            options.followingFrameOptions.frameGeometryAfterManeuver.run {
+                pointsToFrameAfterCurrentStep = getPointsToFrameAfterCurrentManeuver(
+                    frameGeometryAfterManeuverEnabled = enabled,
+                    generatedPostManeuverFramingPoints = postManeuverFramingPoints,
+                    currentLegProgress = currentLegProgress,
+                    currentStepProgress = currentStepProgress
+                )
+            }
+
+            remainingPointsOnRoute = getAllRemainingPointsOnRoute(
+                completeRoutePoints,
+                pointsToFrameOnCurrentStep,
+                currentLegProgress,
+                currentStepProgress
+            )
+        } ?: run {
+            followingPitchProperty.fallback = options.followingFrameOptions.defaultPitch
+            pointsToFrameOnCurrentStep = emptyList()
+            pointsToFrameAfterCurrentStep = emptyList()
+            remainingPointsOnRoute = emptyList()
         }
```

Comment:
Could we extract each of these blocks into `private` `fun`ctions?

Reply:
Same as above.